### PR TITLE
[chore] make graph related attribute lazy

### PR DESCRIPTION
**Pull request by gee-senbong**
## Description

Recent runtime performance investigation found that it takes long time to serialize & deserialize the model when it contains complicated graphs (loading from persistent, convert to response model). To improve the runtime performance, this PR aims to make these attribute lazy (keeps the original dictionary format when it is not loaded, when accessing the attribute, convert them into expected format & then stored it in the model cache).

## Performance Results

With current changes:
* loading save feature list object: reduce from 13 seconds to 1.5 seconds
* create a feature list with 204 saved features & delete the feature list: reduce from 35 seconds to 9-15 seconds

Time to list assets through catalog object:
Before:
```
                     list_method_name                elapsed
5                  list_feature_lists 0 days 00:00:01.198686
7                       list_features 0 days 00:00:00.549711
11                 list_relationships 0 days 00:00:00.060092
13                        list_tables 0 days 00:00:00.027998
3                       list_entities 0 days 00:00:00.014406
6                 list_feature_stores 0 days 00:00:00.012503
1           list_batch_request_tables 0 days 00:00:00.012432
4   list_feature_job_setting_analyses 0 days 00:00:00.012278
0           list_batch_feature_tables 0 days 00:00:00.012140
2                    list_deployments 0 days 00:00:00.011310
8      list_historical_feature_tables 0 days 00:00:00.011035
9             list_observation_tables 0 days 00:00:00.010707
12          list_static_source_tables 0 days 00:00:00.010282
10                list_periodic_tasks 0 days 00:00:00.010119
```

After:
```
                     list_method_name                elapsed
7                       list_features 0 days 00:00:00.207054
5                  list_feature_lists 0 days 00:00:00.146189
11                 list_relationships 0 days 00:00:00.062013
13                        list_tables 0 days 00:00:00.027012
3                       list_entities 0 days 00:00:00.013862
0           list_batch_feature_tables 0 days 00:00:00.012779
6                 list_feature_stores 0 days 00:00:00.012019
8      list_historical_feature_tables 0 days 00:00:00.011921
4   list_feature_job_setting_analyses 0 days 00:00:00.011829
2                    list_deployments 0 days 00:00:00.011525
1           list_batch_request_tables 0 days 00:00:00.011103
10                list_periodic_tasks 0 days 00:00:00.011039
12          list_static_source_tables 0 days 00:00:00.009585
9             list_observation_tables 0 days 00:00:00.009201
```

Time to call list routes:
Before:
```
                            route                elapsed
11                       /feature 0 days 00:00:01.001080
13                  /feature_list 0 days 00:00:00.917046
27                          /task 0 days 00:00:00.059043
15             /feature_namespace 0 days 00:00:00.029259
6                /deployment/all/ 0 days 00:00:00.013126
4                     /credential 0 days 00:00:00.012444
26                         /table 0 days 00:00:00.012057
14        /feature_list_namespace 0 days 00:00:00.011753
22                     /scd_table 0 days 00:00:00.011164
10                   /event_table 0 days 00:00:00.010914
2                        /catalog 0 days 00:00:00.010897
18                    /item_table 0 days 00:00:00.010780
23                      /semantic 0 days 00:00:00.010390
12  /feature_job_setting_analysis 0 days 00:00:00.010327
9                         /entity 0 days 00:00:00.010153
7            /deployment/summary/ 0 days 00:00:00.010127
8                /dimension_table 0 days 00:00:00.010004
0            /batch_feature_table 0 days 00:00:00.009988
17      /historical_feature_table 0 days 00:00:00.009600
21             /relationship_info 0 days 00:00:00.009547
5                     /deployment 0 days 00:00:00.009546
3                        /context 0 days 00:00:00.009056
1            /batch_request_table 0 days 00:00:00.009013
19             /observation_table 0 days 00:00:00.008918
16                 /feature_store 0 days 00:00:00.008862
20                 /periodic_task 0 days 00:00:00.008853
24           /static_source_table 0 days 00:00:00.008725
25                        /status 0 days 00:00:00.004642
```
After:
```
                            route                elapsed
27                          /task 0 days 00:00:00.060037
13                  /feature_list 0 days 00:00:00.057604
11                       /feature 0 days 00:00:00.032201
15             /feature_namespace 0 days 00:00:00.014279
6                /deployment/all/ 0 days 00:00:00.013371
22                     /scd_table 0 days 00:00:00.012458
26                         /table 0 days 00:00:00.011473
18                    /item_table 0 days 00:00:00.011370
4                     /credential 0 days 00:00:00.011175
24           /static_source_table 0 days 00:00:00.011057
5                     /deployment 0 days 00:00:00.010763
14        /feature_list_namespace 0 days 00:00:00.010698
7            /deployment/summary/ 0 days 00:00:00.010665
2                        /catalog 0 days 00:00:00.010518
19             /observation_table 0 days 00:00:00.009953
23                      /semantic 0 days 00:00:00.009901
10                   /event_table 0 days 00:00:00.009800
8                /dimension_table 0 days 00:00:00.009775
1            /batch_request_table 0 days 00:00:00.009714
9                         /entity 0 days 00:00:00.009639
21             /relationship_info 0 days 00:00:00.009638
0            /batch_feature_table 0 days 00:00:00.009490
17      /historical_feature_table 0 days 00:00:00.009405
20                 /periodic_task 0 days 00:00:00.009185
3                        /context 0 days 00:00:00.008861
16                 /feature_store 0 days 00:00:00.008576
12  /feature_job_setting_analysis 0 days 00:00:00.007662
25                        /status 0 days 00:00:00.003958
```

## Type of Change

Label this pull request to place it under the correct category in Release Notes:

|        **Pull Request Label**         | **Category in Release Notes** |
|:-------------------------------------:|:-----------------------------:|
|       `enhancement`, `feature`        |          🚀 Features          |
| `bug`, `refactoring`, `bugfix`, `fix` |    🔧 Fixes & Refactoring     |
|       `build`, `ci`, `testing`        |    📦 Build System & CI/CD    |
|              `breaking`               |      💥 Breaking Changes      |
|            `documentation`            |       📝 Documentation        |
|            `dependencies`             |    ⬆️ Dependencies updates    |

## Checklist

- [ ] I have read the [`CODE_OF_CONDUCT.md`](https://github.com/featurebyte/featurebyte/blob/main/CODE_OF_CONDUCT.md) and [`CONTRIBUTING.md`](https://github.com/featurebyte/featurebyte/blob/main/CONTRIBUTING.md) guides.
- [ ] I have written tests for the changes made.
- [ ] I have written docstrings in [NumpyDoc format](https://numpydoc.readthedocs.io/en/latest/format.html#docstring-standard)
- [ ] I have labeled my Pull Request correctly

### CODE REVIEW

File: featurebyte/models/feature.py [hunk 1]
```diff
@@ -193,14 +193,18 @@ class FeatureModel(FeatureByteCatalogBaseDocumentModel):
...
     node_name: str
     tabular_source: TabularSource = Field(allow_mutation=False)
     readiness: FeatureReadiness = Field(allow_mutation=False, default=FeatureReadiness.DRAFT)
     version: VersionIdentifier = Field(allow_mutation=False, default=None)
     online_enabled: bool = Field(allow_mutation=False, default=False)
     definition: Optional[str] = Field(allow_mutation=False, default=None)
 
+    # special handling for those attributes that are expensive to deserialize
+    # internal_* is used to store the raw data from persistence, _* is used as a cache
+    internal_graph: Any = Field(allow_mutation=False, alias="graph")
```
[the remark this answers is not in the document]
Comment (gee-senbong) [reply featurebyte/models/feature.py:205]: It shouldn't have any backward compatibility issue as the persistent document is still use `graph` (https://github.com/featurebyte/featurebyte/blob/defe49a035477378f33c7de60908134c10a61bae/featurebyte/service/base_document.py#L162).

[the remark this answers is not in the document]
Comment (gee-senbong) [reply featurebyte/models/feature.py:205]: This is similar to `id` attribute (when it is stored in persistent, it stores as `_id`). When loading it back, it becomes `id` attribute.

File: featurebyte/models/feature.py [hunk 2]
```diff
@@ -226,7 +230,7 @@ def _add_derived_attributes(cls, values: dict[str, Any]) -> dict[str, Any]:
         if any(not x for x in derived_attributes):
             # only derive attributes if any of them is missing
             # extract table ids & entity ids from the graph
-            graph = values["graph"]
+            graph = QueryGraph(**values["internal_graph"])
```
Comment (gee-senbong) [inline featurebyte/models/feature.py:233]: When using `root_validator` (`pre=False`), pydantic converts it to `internal_graph`. If `pre=True` is used, the key is still `graph`.

File: featurebyte/service/batch_request_table.py [hunk 3]
```diff
@@ -73,7 +73,7 @@ async def get_batch_request_table_task_payload(
             await self.context_service.get_document(document_id=data.context_id)
 
         return BatchRequestTableTaskPayload(
-            **data.dict(),
+            **data.dict(by_alias=True),
```
Comment (YS-L) [inline featurebyte/service/batch_request_table.py:76]: Curious why is `by_alias` required here and what would be the issue if that is not set? I don't fully follow how it works end to end.

Comment (gee-senbong) [reply featurebyte/service/batch_request_table.py:76]: Basically,
```
feat.dict()                           =>  {...., "internal_graph": [...], ...}
feat.dict(by_alias=True)   =>  {...., "graph": [...], ...}
```

Comment (gee-senbong) [reply featurebyte/service/batch_request_table.py:76]: When the `alias` is used, the constructor parameters expect `alias` but not the attribute value. Outside of the model (like dictionary, json, constructor parameters), we should use alias key. When referring to the attribute, we should use not attribute value. (`root_validator(pre=False)` is a special case here as non-alias key is used inside the `values`).

Comment (YS-L) [reply featurebyte/service/batch_request_table.py:76]: > When the alias is used, the constructor parameters expect alias but not the attribute value.

I see, I wasn't aware of this. Thanks for the explanation!

**Review by YS-L — APPROVED** [review]
The improvement looks great! LGTM

**Pull request merged**